src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+        {
+            uint status = _api.StreamReceiveSetEnabledDelegate(_ptr, enabled: false);
+        }
+
+        internal static uint NativeCallbackHandler(
+           IntPtr stream,
+           IntPtr context,
+           StreamEvent connectionEventStruct)
+        {
+            var handle = GCHandle.FromIntPtr(context);
+            var quicStream = (MsQuicStream)handle.Target;
+
+            return quicStream.HandleEvent(ref connectionEventStruct);
+        }
+
+        private uint HandleEvent(ref StreamEvent evt)
+        {
+            uint status = MsQuicConstants.Success;
+
+            try
+            {
+                switch (evt.Type)
+                {
+                    case QUIC_STREAM_EVENT.START_COMPLETE:
+                        status = HandleStartComplete();
+                        break;
+                    case QUIC_STREAM_EVENT.RECV:
+                        {
+                            HandleEventRecv(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                        {
+                            status = HandleEventSendComplete(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.PEER_SEND_CLOSE:
+                        {
+                            status = HandleEventPeerSendClose();
+                        }
+                        break;
+                    // TODO figure out difference between SEND_ABORT and RECEIVE_ABORT
+                    case QUIC_STREAM_EVENT.PEER_SEND_ABORT:
+                        {
+                            // TODO why is this firing.
+                            status = HandleEventPeerSendAbort();
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.PEER_RECV_ABORT:
+                        {
+                            status = HandleEventPeerRecvAbort();
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventSendShutdownComplete(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                        {
+                            // TODO make sure close is called.
+                            status = HandleEventShutdownComplete();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private void HandleEventRecv(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            static unsafe void CopyToBuffer(Span<byte> buffer, StreamEvent evt)
```

Comment:
I would write this as
```c#
static unsafe void CopyToBuffer(StreamEvent evt, Span<byte> buffer)
{
    Debug.Assert(evt.Data.Recv.BufferCount > 1);

    QuicBuffer evtRecvBuffer = evt.Data.Recv.Buffers[0];

    Debug.Assert(buffer.Length > evtRecvBuffer.Length);

    new Span<byte>(evtRecvBuffer.Buffer, (int)evtRecvBuffer.Length).CopyTo(buffer);
}
```

Or maybe add such a method to `QuicBuffer`-struct?

Reply:
This code seems to assume there will always only be one `Buffers`. That's incorrect.

Reply:
Yeah, I had a feeling this was incorrect. Is it arbitrary when MsQuic would return multiple buffers?